File: hydra-haskell/src/main/haskell/Hydra/Lib/Tuples.hs
-- | Haskell implementations of hydra.lib.tuples primitives

module Hydra.Lib.Tuples where


curry :: ((a, b) -> c) -> a -> b -> c
curry f x y = f (x, y)

fst :: (a, b) -> a
fst = Prelude.fst

snd :: (a, b) -> b
snd = Prelude.snd

uncurry :: (a -> b -> c) -> (a, b) -> c
uncurry f (x, y) = f x y

File: hydra-python/src/main/python/hydra/lib/tuples.py
"""Python implementations of hydra.lib.tuples primitives."""

from collections.abc import Callable
from typing import TypeVar

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')


def curry[A, B, C](f: Callable[[tuple[A, B]], C]) -> Callable[[A], Callable[[B], C]]:
    """Convert a function that takes a pair to a curried function."""
    return lambda x: lambda y: f((x, y))


def fst[A, B](pair: tuple[A, B]) -> A:
    """Get the first element of a pair."""
    return pair[0]


def snd[A, B](pair: tuple[A, B]) -> B:
    """Get the second element of a pair."""
    return pair[1]


def uncurry[A, B, C](f: Callable[[A], Callable[[B], C]]) -> Callable[[tuple[A, B]], C]:
    """Convert a curried function to a function that takes a pair."""
    return lambda pair: f(pair[0])(pair[1])
